wearApp/src/main/java/dev/johnoreilly/confetti/wear/MainActivity.kt

```diff
@@ -57,13 +64,20 @@ class MainActivity : ComponentActivity() {
 
     override fun onNewIntent(intent: Intent?) {
         super.onNewIntent(intent)
-        navController.handleDeepLink(intent)
+
+        if (intent != null) {
+            val handled = appComponent.handleDeeplink(intent)
```

Comment:
I think the deeplink may come in `onCreate` as well? Maybe pass the deeplink via `DefaultWearAppComponent` constructor? And here just call `setIntent` + `recreate`, to go through the constructor.

Reply:
I don't think onCreate is called in this case, only onResume.  I'll change the onCreate constructor call, but think I still need this.

https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)

Reply:
I think there are two scenarios.

1. The activity is already created - the activity instance is reused and the intent is delivered via `onNewIntent`.
2. The activity is not yet created (e.g. the process is killed) - the activity is created and the intent is delivered via `onCreate`.

What do you think?

Reply:
I seem to be getting a third.

The activity is already created - the activity instance is reused and the intent is delivered only via onCreate.

I've tried to follow this but it doesn't appear to work, I see the right intent, but I get the app state from before the deeplink.

```
    override val stack: Value<ChildStack<Config, Child>> =
        childStack(
            source = navigation,
            initialStack = { initialConfig(intent) },
            childFactory = this::buildChild,
        )
```

Reply:
> the activity instance is reused and the intent is delivered only via onCreate

If onCreate is called then it's a new instance of activity. But regardless, there any saved state when this happens? E.g. the savedInstanceState argument is not null? What are the steps to reproduce?

Reply:
Actually it seemed to work.  My repro steps, some of which might show my unfamiliarity with Navigation.

1. Launch confetti, navigate to bookmarks, some non default screen.
2. Close the Wear emulator with 2nd button.
3a. Hit 2nd button again and launch Confetti from All apps view.  Should launch a fresh activity, back to Home screen.  But is showing the bookmarks.
3b. Hit the 1st button and launch Confetti from Recents.  Should reuse the existing activity, working.
3c. Launch from a complication, should launch a fresh activity, to the deeplink.
3d. Launch from a tile, should launch a fresh activity, to the deeplink.

**But don't spend time on it yet, I'll ask at work.**

Reply:
Thanks! I've go some questions.

3a - why launching Confetti from All apps should go to the initial screen? Are there any guidelines?
3c and 3d - what does "to the deeplink" mean? Where the deeplink comes from?

Reply:
The details get into task affinities and task stacks. https://developer.android.com/guide/components/activities/tasks-and-back-stack and https://developer.android.com/guide/components/activities/recents. These are mobile concepts also.

But it's somewhat confusing on Wear even for users. There are different ways to launch apps from the SysUI.

1) The App Launcher, is the top level launches for your App.  you can have multiple e.g. Compose Email, vs see Inbox.

![image](https://github.com/joreilly/Confetti/assets/231923/82318005-228a-4e26-afae-136fac38d936)

It's an alphabetical list of activities on Wear OS.  Samsung shows this as your app icons in a grid 
<img width="147" alt="image" src="https://github.com/joreilly/Confetti/assets/231923/fc6335a8-c255-4180-b694-a5a338b9c37d">

Clicking here should start the thing you clicked, even if it means using an existing activity and changing the navigation stack. 

 2) Recents. This shows what you are currently running.

![image](https://github.com/joreilly/Confetti/assets/231923/b203e4db-7149-426a-8bdf-9f03c61049f2)

Samsung uses a visual display 
<img width="471" alt="image" src="https://github.com/joreilly/Confetti/assets/231923/88727bc2-a820-419a-9c6b-fc6f4f47322d">

Clicking here should take you back to where you were.

Reply:
For complications and tiles.  They should be contextual, so take you straight into a session, but whatever, it should be clear to the user what they will do.

So TaskStackBuilder is used with Androidx Navigation, to restore the synthetic backstack.
![complication](https://github.com/joreilly/Confetti/assets/231923/5b130cc6-f8d5-47f7-a31f-ac5520206c98)

![tile](https://github.com/joreilly/Confetti/assets/231923/98914dd2-c6c2-4a3b-b4b7-456db62fd6e4)

Reply:
Thanks! This seems aligned with mobile phones. I have just played with this [guide](https://developer.android.com/training/app-links/deep-linking), as well as this [doc](https://developer.android.com/reference/android/app/Activity#onNewIntent(android.content.Intent)).

In the manifest, without `android:launchMode="singleTop"`, there is a new instance of activity created for each deep link. Each activity is added to the task stack.

With `android:launchMode="singleTop"`, if there is no created activity, a new one is created (`onCreate` is called), and the deeplink is available via the `Activity#getIntent` method. If there is a created activity, then it gets activated and `onIntent` called with the deeplink in the method's `intent` argument.

In the second case, the ideal approach would be to do the following.

```kotlin
    override fun onNewIntent(intent: Intent) {
        super.onNewIntent(intent)
        setIntent(intent)
        recreate()
    }
```

This recreates the activity and starts it with the deeplink in the intent, as it has not been created before. So you can as usual read the deeplink via `Activity#getIntent` method from `onCreate`. However, this also preserves the activity state, and so the navigation stack is also preserved and the deeplink is not handled. I think currently we should not recreate the activity and just call the component for navigation with the deeplink from `onNewIntent`. I will think how this could be optimized in Decompose.

Reply:
I'll leave as is, but add singleTop.

Thanks

Reply:
Though, I have just tried deeplinks on a normal phone with `singleTop`. When there is an activity already created (e.g. launched from the home screen), and then I click on a deeplink in another app, then a new MainActivity is created and added  to the activity task of the app where I clicked the link. So there are two created activities now.

The behaviour I described in the previous message (with saved state preventing the deeplink from being opened) is reproducible with adb: `./adb shell am start -W -a android.intent.action.VIEW -d "<link>" <pkg>`. Or by using `singleTask` launch mode instead of `singleTop`.

Reply:
Gone with singleTask for now. I'll follow up at work and try to work out how to test for this behaviour.